src/providers/pkcs11_provider/mod.rs

```diff
@@ -0,0 +1,996 @@
+                    }
+                }
+                Err(e) => {
+                    error!("Finding objects failed with {}", e);
+                    Err(ResponseStatus::PsaErrorHardwareFailure)
+                }
+            }
+        }
+    }
+}
+
+impl Provide for Pkcs11Provider {
+    fn list_opcodes(&self, _op: OpListOpcodes) -> Result<ResultListOpcodes> {
+        Ok(ResultListOpcodes {
+            opcodes: SUPPORTED_OPCODES.iter().copied().collect(),
+        })
+    }
+
+    fn describe(&self) -> ProviderInfo {
+        ProviderInfo {
+            // Assigned UUID for this provider: 30e39502-eba6-4d60-a4af-c518b7f5e38f
+            uuid: Uuid::parse_str("30e39502-eba6-4d60-a4af-c518b7f5e38f").unwrap(),
+            description: String::from("PKCS #11 provider, interfacing with a PKCS #11 library."),
+            vendor: String::from("OASIS Standard."),
+            version_maj: 0,
+            version_min: 1,
+            version_rev: 0,
+            id: ProviderID::Pkcs11Provider,
+        }
+    }
+
+    fn create_key(&self, app_name: ApplicationName, op: OpCreateKey) -> Result<ResultCreateKey> {
+        info!("Pkcs11 Provider - Create Key");
+
+        if op.key_attributes.key_type != KeyType::RsaKeypair
+            || op.key_attributes.algorithm != Algorithm::sign(SignAlgorithm::RsaPkcs1v15Sign, None)
+        {
+            error!("The PKCS 11 provider currently only supports creating RSA key pairs for signing and verifying.");
+            return Err(ResponseStatus::UnsupportedOperation);
+        }
+
+        let key_name = op.key_name;
+        // This should never panic on 32 bits or more machines.
+        let key_size = std::convert::TryFrom::try_from(op.key_attributes.key_size).unwrap();
+
+        let key_triple = KeyTriple::new(app_name, ProviderID::Pkcs11Provider, key_name);
+        let mut store_handle = self.key_id_store.write().expect("Key store lock poisoned");
+        let mut local_ids_handle = self.local_ids.write().expect("Local ID lock poisoned");
+        if key_id_exists(&key_triple, &*store_handle)? {
+            return Err(ResponseStatus::KeyAlreadyExists);
+        }
+        let key_id = create_key_id(
+            key_triple.clone(),
+            &mut *store_handle,
+            &mut local_ids_handle,
+        )?;
+
+        let mech = CK_MECHANISM {
+            mechanism: pkcs11::types::CKM_RSA_PKCS_KEY_PAIR_GEN,
+            pParameter: std::ptr::null_mut(),
+            ulParameterLen: 0,
+        };
+
+        let mut priv_template: Vec<CK_ATTRIBUTE> = Vec::new();
+        let mut pub_template: Vec<CK_ATTRIBUTE> = Vec::new();
+
+        priv_template
+            .push(CK_ATTRIBUTE::new(pkcs11::types::CKA_SIGN).with_bool(&pkcs11::types::CK_TRUE));
+        priv_template.push(CK_ATTRIBUTE::new(pkcs11::types::CKA_ID).with_bytes(&key_id));
+        priv_template
+            .push(CK_ATTRIBUTE::new(pkcs11::types::CKA_TOKEN).with_bool(&pkcs11::types::CK_TRUE));
+
+        pub_template
```

Comment:
Good point! I checked that the implementation I was working on (SoftHSM) does that by default but it would be more correct indeed to fix it to `0x010001`. Will change.